### PR TITLE
Error reporting added!

### PR DESCRIPTION
**Pull request by mistergone**
This PR adds error handling. Now, when numbers exceed limits, the object returned by `student-debt-calc` contains an error property, which is itself an object full of error keys and messages.
## Additions
- `financials` object returned by this package has an `errors` property (an Object). The `errors` Object starts off empty, and as errors occur, it is populated with keys (error codes) and values (error messages). For instance, if the passed values represent a Pell grant that exceeds a federal limit of $5,750, the `errors` object would be: `{ 'pellOverCap': 'The Pell grant exceeds the federal limit of 5750.' }`. In this way, an app developer can check to see if the error code is an extant property of the `errors` Object (e.g. `financials.errors.hasOwnProperty( 'someErrorCode' )` ), and can use the default error message if they wish.
## Changes
- Some aspects of the code were altered slightly to make error reporting more simple.
## Testing
- `mocha test/` runs our suite of tests, including 9 new tests for error reporting.
## Review
- @ascott1, if you have the time!
- @marteki, @niqjohnson, check it out!
## Checklist
- [x] Changes are limited to a single goal (no scope creep)
- [x] Code can be automatically merged (no conflicts)
- [x] Code follows the standards laid out in the [front end playbook](https://github.com/cfpb/front-end)
- [x] Passes all existing automated tests
- [x] New functions include new tests
- [x] New functions are documented (with a description, list of inputs, and expected output)
- [x] Placeholder code is flagged
- [x] Project documentation has been updated (including the "Unreleased" section of the CHANGELOG)
## Animated GIF

liz-lemon-speak-to-pizza

### CODE REVIEW

File: src/loans/direct-subsidized.js [hunk 1]
```diff
@@ -18,16 +18,34 @@ function subDirect( data ) {
...
   }
 
+  // Error handling
+  if ( data.directSubsidized > 0 && data.undergrad === false ) {
+    data.errors.subsidizedNoGrad = 'Direct subsidized loans are available only to undergraduate students.';
+  }
+  if ( data.directSubsidized > data.yearOneCosts - data.pell - data.perkins ) {
+    data.errors.subsidizedOverCost = 'Direct subsidized loans exceed cost of attendance.';
+  }
+  console.log( data.directSubsidizedMax );
```
Comment (niqjohnson) [inline src/loans/direct-subsidized.js]: Remove this?

**Comment by niqjohnson** [comment]
Other than the line note above, looks :+1: to me. This is going to be a huge help with showing these errors in the college-costs interface.

**Comment by mistergone** [comment]
`console.log()` of shame and some other linting errors were cleaned up!

**Comment by ascott1** [comment]
@mistergone I like what you're doing here and the idea of taking care of edge cases and giving something back that is useful to both users and module consumers. So, overall :+1: to the specific errors you're catching and the ideas presented here.

Typically, I would expect an error case to specifically not allow an action to happen or throw an error `throw new Error('something bad happened');` (though I don't think `throw` would be useful for you here).

Are those error messages being used by the UI? Could come of the logic or at least the specific language be handled in the app rather than handling it here?

I want to avoid a scenario where we're too tightly coupling this and the UI. That way if a wording change needs to happen it happens there rather than here. I could be misunderstanding the use case though.

**Comment by mistergone** [comment]
@ascott1 Great questions!

**First**, on the nature of "errors" - these errors aren't the sort of "fatal errors" or "exceptions" that might come from a piece of software, but rather "errors" where a user has input a value that is incorrect. For instance, the federal government sets limits on Pell grants, but generally, the package wants to accept those values, limit the values, and report a message about the fact that it limited those values.

Maybe the word "error" is the wrong word to use because of its technical definition...

Anyway, we generally don't want the calculator to stop working in these cases - we want it to assume the user was unaware of the limit, and notify them of the error.

The package also allows the developer to pass new values for these limits, keeping the application evergreen or adaptable for custom or future-leaning scenarios.

**Second**, the interaction of a given UI with these errors - I _definitely_ thought of this, which is why I used the concept of "error codes" instead of just returning error messages. However, since the error codes were keys in an object, I figured I could use the value of each key as a plain language explanation of the error.

Does this tie them to closely to our UI? Not at all, in my opinion. In fact, I highly doubt we'll use the error messages I included in the code (I didn't run them by any content or UX person, for instance!). Instead, we'll just check for the presence of error codes as keys in the `financials.errors` Object, and we'll show or hide UI elements which contain our carefully crafted content. The messages are purely there as a little bit of convenience for developers who might use this package.

In this way, all content changes can happen in the application (for instance, in our case, college-costs) because the messages themselves can be ignored in favor of simply checking which error codes (keys in `financials.errors`) exist.

Here's an example of what I see as this being used in front-end code:

```
// Example errors Object after student-debt-calc has run
financials.errors = {
  'pellOverCosts': 'The Pell grant exceeds the cost of attendance.' ,
  'mtaOverCap': 'Military Tuition Assistance exceeds the federal limit of 4500'
};

// We can simply check to see what keys exist, and then show() elements based
// on those keys...
for ( errorCode in financials.errors ) {
  $( '[data-error-code="' + errorCode + '"]' ).show();
}
```

That's overly simplistic, but it's the basic idea of how I would recommend a developer implementing these "errors" in an application.

This is a good note because I should write some better documentation to explain little things about this package such as my intent with the `errors` Object.

**Comment by ascott1** [comment]
That makes sense to me @mistergone. In that case, I say let's merge, but add some documentation around what you just explained so future peeps understand the intent as well.

**Comment by mistergone** [comment]
Grrrrrreat!

tony-the-tiger

**Pull request merged**